syntax = "proto3";

import "app/models/proto/shared.proto";
import "buf/validate/validate.proto";

service ChangesetService {
  rpc GetChangeset(GetChangesetRequest) returns (GetChangesetResponse) {
    option idempotency_level = NO_SIDE_EFFECTS;
  }
}

message GetChangesetRequest {
  uint64 id = 1 [(buf.validate.field).uint64.gt = 0];
}

message GetChangesetResponse {
  ChangesetData changeset = 1 [(buf.validate.field).required = true];
}

message ChangesetData {
  message Element {
    uint64 id = 1;
    uint64 version = 2;
    bool visible = 3;
    optional string name = 4;
    optional ElementIcon icon = 5;
  }

  uint64 id = 1;
  optional User user = 2;
  uint64 created_at = 3;
  optional uint64 closed_at = 4;
  uint32 num_create = 5;
  uint32 num_modify = 6;
  uint32 num_delete = 7;
  string comment_rich = 8;
  map<string, string> tags = 9;
  repeated Bounds bounds = 10;
  repeated Element nodes = 11;
  repeated Element ways = 12;
  repeated Element relations = 13;
  optional uint64 prev_changeset_id = 14;
  optional uint64 next_changeset_id = 15;
  bool is_subscribed = 16;
}

message ChangesetCommentPage {
  message Comment {
    User user = 1;
    uint64 created_at = 2;
    string body_rich = 3;
  }

  repeated Comment comments = 1;
}

message ChangesetCommentResult {
  ChangesetData changeset = 1 [(buf.validate.field).required = true];
  ChangesetCommentPage comments = 2 [(buf.validate.field).required = true];
}

message RenderChangesetsData {
  message Changeset {
    uint64 id = 1;
    optional User user = 2;
    repeated Bounds bounds = 3;
    bool closed = 4;
    uint64 status_changed_at = 5;
    optional string comment = 6;
    uint32 num_create = 7;
    uint32 num_modify = 8;
    uint32 num_delete = 9;
    uint64 num_comments = 10;
  }

  repeated Changeset changesets = 1;
}
